### ISSUE TITLE
Provide a callback function to handle serious errors

### ISSUE THREAD

**Issue by benjie** [graphile/postgraphile#1073]
Rather than logging/exiting, allow the application to handle the fail. This needs to give the application a way of retrying (e.g. a retry function they can call after a delay) should they wish to, which also allows them to override the backoff algorithm.

**Comment by pyramation** [comment]
any workarounds for this? maybe an error handler that we can pass into the config? if the handler is specified, the app won't just crash or keep trying, the app can sort out depending on the use case (similar to https://github.com/graphile/postgraphile/issues/1042)

**Comment by pyramation** [comment]
why not just throw this? https://github.com/graphile/postgraphile/blob/master/packages/postgraphile/src/postgraphile/postgraphile.ts#L141

**Comment by pyramation** [comment]
also, can you explain what use cases are for the `attempts`? https://github.com/graphile/postgraphile/blob/master/packages/postgraphile/src/postgraphile/postgraphile.ts#L137

shouldn't this be handled at an infrastructure level?

**Comment by pyramation** [comment]
Here's a quick attempt at an idea: https://github.com/graphile/postgraphile/pull/1302/files

caveat: To avoid the typescript yelling for returning in the while loop, I just (for proof of concept) had the function return a GraphQLSchema, but I know that's incorrect.

**Comment by pyramation** [comment]
should I be looking in `postgraphile-core`?

**Comment by benjie** [comment]
The error is generated asynchronously so it couldn’t be caught in that try/catch - the code has already moved on.

**Comment by pyramation** [comment]
so for v5, do you plan to do this retry logic? 

I would suggest moving it to the infrastructure level, and I'd even help prepare an example for you to see how to decouple it from a devops perspective :) let me know and I can start sending some example repos

**Comment by pyramation** [comment]
also, would you be open to merging a PR or helping me do this PR for v4? My code in my example does what I need from an ops perspective. I just couldn't get the Typescript to be perfect since the function clearly doesn't return a schema. Potentially it may be better to make an option that doesn't call the `exit()`?

Folks using k8s, docker, or anyone running more than one instance of graphile would run into this issue, and the PR as an example I made does the trick, although not perfectly

**Comment by pyramation** [comment]
updated the PR: https://github.com/graphile/postgraphile/pull/1302

I cannot see why the test has failed.. the app it's connected to won't show me anything. Happy to fix

**Comment by benjie** [comment]
Discussion in #1302

**Comment by benjie** [comment]
Current proposed fix for this: https://github.com/graphile/postgraphile/pull/1342

**Comment by pyramation** [comment]
> Current proposed fix for this: #1342

I can say this works well!

what's better?

```js
    async retryOnInitFail(error) {
      throw error;
    },
```

or 

```js
    async retryOnInitFail(error) {
      return false;
    },
```

They seem to have the same behavior (which both work for me!) thanks again :)

**Comment by benjie** [comment]
Thanks for confirmation, happy to hear it! They're equivalent :+1:
